### PR TITLE
Allow vehicle to receive a command_ack &  Safely avoid send the same vehicle_command/command_ack back

### PR DESCRIPTION
**Pull request by zehortigoza**
This patches are part of the changes that need to be done to implement request of arm authorization (https://github.com/PX4/Firmware/pull/7377) sending it apart to get better and faster reviews.

### CODE REVIEW

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -630,6 +630,21 @@ MavlinkReceiver::handle_message_command_ack(mavlink_message_t *msg)
...
+
+	if (_command_ack_pub == nullptr) {
+		_command_ack_pub = orb_advertise_queue(ORB_ID(vehicle_command_ack), &command_ack,
+						       vehicle_command_ack_s::ORB_QUEUE_LENGTH);
+
+	} else {
+		orb_publish(ORB_ID(vehicle_command_ack), _command_ack_pub, &command_ack);
+	}
+
+	// TODO: move it to the same place that sent the command
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp]: So this should presumably be moved to the camera_trigger?

File: msg/vehicle_command.msg [hunk 2]
```diff
@@ -95,12 +95,13 @@ float32 param1			# Parameter 1, as defined by MAVLink uint32 VEHICLE_CMD enum.
 float32 param2			# Parameter 2, as defined by MAVLink uint32 VEHICLE_CMD enum.
 float32 param3			# Parameter 3, as defined by MAVLink uint32 VEHICLE_CMD enum.
 float32 param4			# Parameter 4, as defined by MAVLink uint32 VEHICLE_CMD enum.
-float64 param5			# Parameter 5, as defined by MAVLink uint32 VEHICLE_CMD enum.
-float64 param6			# Parameter 6, as defined by MAVLink uint32 VEHICLE_CMD enum.
+float32 param5			# Parameter 5, as defined by MAVLink uint32 VEHICLE_CMD enum.
```
Comment (julianoes) [inline msg/vehicle_command.msg]: The point is that if we handle lat or lon in these params, we need the accuracy of double.

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 1 above)
Comment (zehortigoza) [inline src/modules/mavlink/mavlink_receiver.cpp]: yes

File: msg/vehicle_command.msg (hunk 2 above)
Comment (zehortigoza) [inline msg/vehicle_command.msg]: Ok, I fixed this by doing the division on the right place and I also check all other users of param5 and param6 and no other needs more precision than float.

File: src/modules/navigator/navigator_main.cpp [hunk 3]
```diff
@@ -459,8 +467,16 @@ Navigator::task_main()
 				}
 
 				if (PX4_ISFINITE(cmd.param5) && PX4_ISFINITE(cmd.param6)) {
-					rep->current.lat = (cmd.param5 < 1000) ? cmd.param5 : cmd.param5 / (double)1e7;
-					rep->current.lon = (cmd.param6 < 1000) ? cmd.param6 : cmd.param6 / (double)1e7;
+					rep->current.lat = cmd.param5;
+					rep->current.lon = cmd.param6;
+
+					if (rep->current.lat < 1000) {
+						rep->current.lat /= (double)1e7;
```
Comment (julianoes) [inline src/modules/navigator/navigator_main.cpp]: Oh, now I see that you actually inverted the logic here. This is wrong now.

File: msg/vehicle_command.msg (hunk 2 above)
Comment (julianoes) [inline msg/vehicle_command.msg]: I don't think it matters where you do the division. If you store lat/lon in a float, you lose accuracy.

File: src/modules/navigator/navigator_main.cpp (hunk 3 above)
Comment (zehortigoza) [inline src/modules/navigator/navigator_main.cpp]: Ohh my bad fixed...

File: msg/vehicle_command.msg (hunk 2 above)
Comment (zehortigoza) [inline msg/vehicle_command.msg]: The MAVLink message uses a float so there is no loss in here, what I did was to set the float to the internal double so all the calculations will use double precision.

Comment (julianoes) [inline msg/vehicle_command.msg]: Ok, fair enough. I was thinking of the case where you have a command int with type `int32_t`.

Comment (LorenzMeier) [inline msg/vehicle_command.msg]: @zehortigoza Please change back to double. As Julian says this is causing a loss of precision for any command containing GPS coordinates by about 1.6 meters.

Comment (zehortigoza) [inline msg/vehicle_command.msg]: @LorenzMeier There is no MAVLink message with double, so there is no loss in here.

File: src/drivers/vmount/output_mavlink.cpp [hunk 4]
```diff
@@ -54,7 +54,7 @@ OutputMavlink::OutputMavlink(const OutputConfig &output_config)
 
 int OutputMavlink::update(const ControlData *control_data)
 {
-	vehicle_command_s vehicle_command;
+	vehicle_command_s vehicle_command = {};
```
Comment (LorenzMeier) [inline src/drivers/vmount/output_mavlink.cpp]: This is forcing an initialisation on every iteration - is this really required? Benchmarking PX4 we found that the system is spending quite a bit of time in these defensive programming patterns. Which are generally good, but induce a lot of load (5-8% CPU IIRC) in embedded.

File: src/drivers/vmount/input_mavlink.cpp [hunk 5]
```diff
@@ -285,7 +285,7 @@ int InputMavlinkCmdMount::update_impl(unsigned int timeout_ms, ControlData **con
 
 void InputMavlinkCmdMount::_ack_vehicle_command(uint16_t command)
 {
-	vehicle_command_ack_s vehicle_command_ack;
+	vehicle_command_ack_s vehicle_command_ack = {};
```
Comment (LorenzMeier) [inline src/drivers/vmount/input_mavlink.cpp]: This is forcing an initialisation on every iteration - is this really required? Benchmarking PX4 we found that the system is spending quite a bit of time in these defensive programming patterns. Which are generally good, but induce a lot of load (5-8% CPU IIRC) in embedded.

File: src/modules/sdlog2/sdlog2.c [hunk 6]
```diff
@@ -385,6 +385,8 @@ int sdlog2_main(int argc, char *argv[])
 
 	if (!strncmp(argv[1], "on", 2)) {
 		struct vehicle_command_s cmd;
+
+		memset(&cmd, 0, sizeof(cmd));
```
Comment (LorenzMeier) [inline src/modules/sdlog2/sdlog2.c]: Zeroing here makes sense, no runtime expense.

File: src/drivers/vmount/output_mavlink.cpp (hunk 4 above)
Comment (dagar) [inline src/drivers/vmount/output_mavlink.cpp]: You can do it like this. 
https://github.com/PX4/Firmware/blob/master/src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp#L463

File: src/drivers/vmount/input_mavlink.cpp (hunk 5 above)
Comment (zehortigoza) [inline src/drivers/vmount/input_mavlink.cpp]: As in further patch in this PR a new field is added to vehicle_command_ack I choose to zeroing all vehicle_command_ack to make it safe but I can just set this field in all vehicle_command_ack.

File: src/drivers/vmount/output_mavlink.cpp (hunk 4 above)
Comment (zehortigoza) [inline src/drivers/vmount/output_mavlink.cpp]: As in further patch in this PR a new field is added to vehicle_command I choose to zeroing all vehicle_command to make it safe but I can just set this field in all vehicle_command.

File: src/modules/sdlog2/sdlog2.c (hunk 6 above)
Comment (zehortigoza) [inline src/modules/sdlog2/sdlog2.c]: As in further patch in this PR a new field is added to vehicle_command I choose to zeroing all vehicle_command to make it safe but I can just set this field in all vehicle_command.

File: src/drivers/vmount/output_mavlink.cpp (hunk 4 above)
Comment (LorenzMeier) [inline src/drivers/vmount/output_mavlink.cpp]: See the hint from @dagar how to not miss fields but avoid the overhead.

**Comment by LorenzMeier** [comment]
@zehortigoza I've restarted Circle CI.

File: src/drivers/vmount/output_mavlink.cpp (hunk 4 above)
Comment (zehortigoza) [inline src/drivers/vmount/output_mavlink.cpp]: I can do that but keep in mind that this type of initialization + code generator that do struct alignment could cause several errors like this:
```px4-firmware/src/modules/commander/commander.cpp:4098:2:` sorry, unimplemented: non-trivial designated initializers not `supported```
if one variable is added to the struct and it changes the order.

Comment (zehortigoza) [inline src/drivers/vmount/output_mavlink.cpp]: Changes done

**Comment by LorenzMeier** [comment]
Rebased here: https://github.com/PX4/Firmware/pull/7732

**Pull request closed without merging**